### ISSUE TITLE
Package naming policies for Julia packages

### ISSUE THREAD

**Issue by mkitti** [conda-forge/conda-forge.github.io#1822]
### Your question:

# Question

How should Julia packages be named within conda-forge?

# Introduction

The [julia-feedstock](https://github.com/conda-forge/julia-feedstock) is now current with upstream Julia at version 1.8.0. Support is currently best on Linux. Mac OS X support is possible on x86_64 and arm64 but needs further work and testing. Windows support is likely not possible within conda-forge without a comprehensive mingw-w64 ecosystem as opposed to a Microsoft Visual C.

We current have two feedstocks employing a dual language Python - Julia strategy within conda-forge:
* https://github.com/ngam/pysr-feedstock
* https://github.com/conda-forge/xbitinfo-feedstock

These two packages primarily are Python interfaces for the following Julia packages:
* [SymbolicRegression.jl](https://github.com/MilesCranmer/SymbolicRegression.jl)
* [BitInformation.jl](https://github.com/milankl/BitInformation.jl)

Currently the installation of the Julia packages needs to be initiated by the user and is handled completely outside of the `conda` or `mamba` package management. The Julia feedstock configures the Julia depot within the conda prefix rather than the default `~/.julia` depot location.

It would best if these Julia packages could be directly installed by `conda` rather than having the user initiate a manual step. I have [developed a mechanism using the Julia depot stack that allows `conda` to directly install Julia packages into the user's Julia load path](https://github.com/conda-forge/pysr-feedstock/pull/43#issuecomment-1233129264) if they use a preconfigured Julia project environment. Thus we are now prepared to propose conda-forge feedstocks for individual Julia packages and now need to decide upon a naming scheme.

# Julia package naming guidelines

Julia packages typically include capitalization according to camel case and a "file extension" with a preceding period. For example:

* [PyCall.jl
](https://github.com/JuliaPy/PyCall.jl)
* [Conda.jl](https://github.com/JuliaPy/Conda.jl)
* [HDF5.jl](https://github.com/JuliaIO/HDF5.jl)
* [HDF5_jll.jl](https://github.com/JuliaBinaryWrappers/HDF5_jll.jl)
* [ClusterManagers.jl](https://github.com/JuliaParallel/ClusterManagers.jl)
* [SymbolicRegression.jl](https://github.com/MilesCranmer/SymbolicRegression.jl)
* [BitInformation.jl](https://github.com/milankl/BitInformation.jl)

Julia package naming guidelines are located at the following URL:
https://pkgdocs.julialang.org/v1/creating-packages/#Package-naming-guidelines-1

# Potential naming schemes within conda-forge

My first preference would be to preserve the original package names. Julia packages can be uniquely identified by the ".jl" extension from those in other languages.

From #18, many seem to prefer no capitalization. If we do want to conform to a language prefix with all lowercase, my preference would be "julia-" since this would unambiguous:
* julia-pycall
* julia-conda
* julia-hdf5
* julia-hdf5_jll
* julia-clustermanagers
* julia-symbolicregression
* julia-bitinformation

I could also see an argument for using "jl-" based on brevity and similarity to the original name:
* jl-pycall
* jl-conda
* jl-hdf5
* jl-hdf5_jll
* jl-clustermanagers
* jl-symbolicregression
* jl-bitinformation

cc: @ngam, @isuruf @ocefpaf

# References

1. https://github.com/conda-forge/julia-feedstock/issues/14
2. https://github.com/conda-forge/pysr-feedstock/pull/43#issuecomment-1233129264

**Comment by mkitti** [comment]
cc: @MilesCranmer, @observingClouds, @rsignell-usgs

**Comment by MilesCranmer** [comment]
`julia-*` gets a +1 from me as well.

**Comment by mkitti** [comment]
Here's yet another [alternative proposal:](https://discourse.julialang.org/t/package-naming-policies-for-julia-packages-within-conda-forge/86659/6?u=mkitti)

* py_call-jl
* conda-jl
* hdf5-jl
* hdf5_jll-jl
* cluster_managers-jl
* symbolic_regression-jl
* bit_information-jl

The `-jl` suffix does preserve the original name a bit more. I'm not so sure if converting CamelCase to snake_case makes sense though. Is there any precedent for snake_case in conda-forge?

**Comment by carterbox** [comment]
> The -jl suffix does preserve the original name a bit more. I'm not so sure if converting CamelCase to snake_case makes sense though. Is there any precedent for snake_case in conda-forge?

Typically, package names are converted to all lower case using the jinja template `{{ name|lower }}`, so there are no CamelCase package names. Some packages use snake_case in order to match the python module name, but most packages are kebab-case.

**Comment by mkitti** [comment]
My understanding from meeting with conda-forge core is that `.` is allowed so we could proceed with lower case names as follows:

* pycall.jl
* conda.jl
* hdf5.jl
* hdf5_jll.jl
* clustermanagers.jl
* symbolicregression.jl
* bitinformation.jl

**Comment by MilesCranmer** [comment]
I think the Julia community would be happier with that - which is a good thing (!) for making them open to an alternative package management system to `Pkg`

**Comment by MilesCranmer** [comment]
> My understanding has always been that `.jl` was just to distinguish julia packages on github, but nothing to do with the actual packages per se.

Good point, I take back my comment. `julia-*` is still my vote then.

**Comment by mkitti** [comment]
In the Julia package manager, you can actually type `add SymbolicRegression.jl` and it will work. In that context, it's just superfluous though.

```
(jl_IyuX1c) pkg> add SymbolicRegression.jl
   Resolving package versions...
  No Changes to `/tmp/jl_IyuX1c/Project.toml`
  No Changes to `/tmp/jl_IyuX1c/Manifest.toml`

```

The other thing I'm noting is that the conda does not seem to care about capitalization. I could spell it `NumPy` and conda will install the `numpy` package. 

```
$ conda create -n test -c conda-forge NumPy
```

This means I could do

```
$ conda create -n test -c conda-forge SymbolicRegression.jl
```

and conda will install the package `symbolicregression.jl`.

**Comment by mkitti** [comment]
A cross reference:
https://github.com/JuliaLang/METADATA.jl/issues/9480

**Comment by mkitti** [comment]
My first preference is `*.jl`. My second preference is `julia-*`. `jl-*` does not seem that useful to me.

**Comment by mkitti** [comment]
Hmm, I guess julia-symbolicregression-feedstock kind of make sense

**Comment by mkitti** [comment]
What is the mechanism to resolve a feedstock name from a package name?

**Comment by mkitti** [comment]
Is there a way to differentiate if a package has been explicitly installed in a conda environment rather than installed because it is a dependency of another package that was requested?

**Comment by cjdoris** [comment]
I'm fairly ambivalent, so long as we're consistent and cannot clash with anything that already exists. I mildly prefer `julia-pycall` (as it's really clear) then `pycall.jl` (=`PyCall.jl` since package names are case insensitive).

**Comment by mkitti** [comment]
I think ngam is suggesting we have both `julia-pycall` and `pycall.jl` with one of those being an alias. 

`conda install -c conda-forge PyCall.jl`

One reason I'm interested in calling the package `PyCall.jl` is search engine optimization. If you searched for "PyCall.jl", `julia-pycall` might not come up in the results. One might not realize there is a conda-forge package for it.

**Comment by cjdoris** [comment]
SEO is a strong argument, though you'd hope that if we put "PyCall.jl" in the description of the Conda package (a description policy would be good too) then that would suffice.

**Comment by majidaldo** [comment]
bump.

**Comment by mkitti** [comment]
Hi @majidaldo , do you have a specific application in mind?

**Comment by mkitti** [comment]
I will try to reengage with this once Julia 1.9 is released, which is imminent. There are some interesting possibilities there for conda-forge such as distributing precompiled binary shared libraries.

cc: @cjdoris

**Comment by majidaldo** [comment]
> Hi @majidaldo , do you have a specific application in mind?

No. But as a polyglot, I just don't want to deal with each programming language's packaging and have to deal with an integration mess as a result.

It should be in the interest of the julia team to work /with/ conda while still keeping their specific way of doing things as has been done with python and R.

multilanguage is the future (already here?) and conda is the only way.

**Comment by mkitti** [comment]
> It should be in the interest of the julia team to work /with/ conda while still keeping their specific way of doing things as has been done with python and R.

The "julia team" here is basically me with some help from @ngam. I do not expect the core Julia developers to responsible for building Julia here nor do I expect CPython developers to be building CPython here either.

We currently have one prototype package online.

https://anaconda.org/conda-forge/versionparsing.jl

A couple others were staged but stalled.

Parsers.jl: https://github.com/conda-forge/staged-recipes/pull/20655
MacroTools.jl: https://github.com/conda-forge/staged-recipes/pull/20654

We also need to resolve what to do about `pyjuliacall`

https://github.com/conda-forge/staged-recipes/pull/20379

If you are interested in volunteering, let me know. There are a lot of issues to resolve.
